server/src/main/java/org/elasticsearch/monitor/fs/FsHealthService.java

```diff
@@ -0,0 +1,197 @@
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.LongSupplier;
+
+/**
+ * Runs periodically and attempts to create a temp file to see if the filesystem is writable. If not then it marks the
+ * path as unhealthy.
+ */
+public class FsHealthService extends AbstractLifecycleComponent {
+
+    private static final Logger logger = LogManager.getLogger(FsHealthService.class);
+
+    private final ThreadPool threadPool;
+    private volatile boolean enabled;
+    private volatile TimeValue refreshInterval;
+    private volatile TimeValue healthCheckTimeoutInterval;
+    private final NodeEnvironment nodeEnv;
+    private final LongSupplier currentTimeMillisSupplier;
+    private Map<Path, TimeStampedStatus> pathHealthStats;
+    private volatile Set<Scheduler.Cancellable> scheduledFutures;
+
+    enum Status { HEALTHY, UNHEALTHY, UNKNOWN }
+
+    public static final Setting<Boolean> ENABLED_SETTING =
+        Setting.boolSetting("monitor.fs.health.enabled", true, Setting.Property.NodeScope, Setting.Property.Dynamic);
+    public static final Setting<TimeValue> REFRESH_INTERVAL_SETTING =
+        Setting.timeSetting("monitor.fs.health.refresh_interval", TimeValue.timeValueSeconds(1), TimeValue.timeValueSeconds(1),
+            Setting.Property.NodeScope);
+    public static final Setting<TimeValue> HEALTHY_TIMEOUT_SETTING =
+        Setting.timeSetting("monitor.fs.health.healthy_timeout", TimeValue.timeValueSeconds(1), TimeValue.timeValueSeconds(1),
+            Setting.Property.NodeScope, Setting.Property.Dynamic);
+
+    FsHealthService(Settings settings, ClusterSettings clusterSettings, ThreadPool threadPool, NodeEnvironment nodeEnv,
+                    LongSupplier currentTimeMillisSupplier) {
+        this.scheduledFutures = new HashSet<>();
+        this.threadPool = threadPool;
+        this.enabled = ENABLED_SETTING.get(settings);
+        this.refreshInterval = REFRESH_INTERVAL_SETTING.get(settings);
+        this.healthCheckTimeoutInterval = HEALTHY_TIMEOUT_SETTING.get(settings);
+        this.currentTimeMillisSupplier = currentTimeMillisSupplier;
+        this.nodeEnv = nodeEnv;
+        this.pathHealthStats = new HashMap<>();
+        clusterSettings.addSettingsUpdateConsumer(HEALTHY_TIMEOUT_SETTING, this::setHealthCheckTimeoutInterval);
+        clusterSettings.addSettingsUpdateConsumer(ENABLED_SETTING, this::setEnabled);
+    }
+
+    @Override
+    protected void doStart() {
+        for (Path path : nodeEnv.nodeDataPaths()) {
+            scheduledFutures.add(threadPool.scheduleWithFixedDelay(new FsPathHealthMonitor(path), refreshInterval,
```

Comment:
The idea behind it is there is still a possibility that multiple data paths can  be mounted on separate (network) volumes and can fail independently. Since we were publishing stats per data path, it made sense to individually report the health per data path. Let me know if you think otherwise. If you don't think `/_node/_local/stats` adds any value we can consider alternatives

Reply:
It's true that the paths can fail independently but this doesn't matter, we will fail the node if any of the paths are broken. I see that independent checks may be useful for stats but as per my previous comment I don't think we need to expose this in stats.